src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
+        } catch (e) {
+            throw new S3Error(`Failed to find region for bucket ${bucketName}: ${e}`)
+        }
+    }
+}
+
+export interface FileStreams {
+    createReadStream(uri: vscode.Uri): stream.Readable
+    createWriteStream(uri: vscode.Uri): stream.Writable
+}
+
+export class DefaultBucket implements Bucket {
+    public readonly name: string
+    public readonly region: string
+    public readonly arn: string
+
+    public constructor({ partitionId, region, name }: { partitionId: string; region: string; name: string }) {
+        this.name = name
+        this.region = region
+        this.arn = `arn:${partitionId}:s3:::${name}`
+    }
+
+    public toString(): string {
+        return `Bucket (name=${this.name}, region=${this.region}, arn=${this.arn})`
+    }
+}
+
+export class DefaultFolder implements Folder {
+    public readonly name: string
+    public readonly path: string
+    public readonly arn: string
+
+    public constructor({ partitionId, bucketName, path }: { partitionId: string; bucketName: string; path: string }) {
+        this.path = path
+        this.arn = `arn:${partitionId}:s3:::${bucketName}/${path}`
+        this.name = _(this.path)
+            .split(DEFAULT_DELIMITER)
+            .dropRight()!
+            .last()!
+    }
+
+    public toString(): string {
+        return `Folder (name=${this.name}, path=${this.path}, arn=${this.arn})`
+    }
+}
+
+export class DefaultFile implements File {
+    public readonly name: string
+    public readonly key: string
+    public readonly arn: string
+    public readonly lastModified: Date | undefined
+    public readonly sizeBytes: number | undefined
+
+    public constructor({
+        partitionId,
+        bucketName,
+        key,
+        lastModified,
+        sizeBytes,
+    }: {
+        partitionId: string
+        bucketName: string
+        key: string
+        lastModified: Date | undefined
+        sizeBytes: number | undefined
+    }) {
+        this.name = _(key)
+            .split(DEFAULT_DELIMITER)
+            .last()!
+        this.key = key
+        this.arn = `arn:${partitionId}:s3:::${bucketName}/${key}`
+        this.lastModified = lastModified
+        this.sizeBytes = sizeBytes
+    }
+
+    public toString(): string {
+        return `File (name=${this.name}, key=${this.key}, arn=${this.arn}, lastModified=${this.lastModified}, sizeBytes=${this.sizeBytes})`
+    }
+}
+
+class DefaultFileStreams implements FileStreams {
+    public createReadStream(uri: vscode.Uri): stream.Readable {
+        return fs.createReadStream(uri.fsPath)
+    }
+
+    public createWriteStream(uri: vscode.Uri): stream.Writable {
+        return fs.createWriteStream(uri.fsPath)
+    }
+}
+
+async function createSdkClient(regionCode: string): Promise<S3> {
```

Comment:
Our other client implementations put this in the class. Any reason to not have it there?

Reply:
Yes, `TS2333: 'this' cannot be referenced in constructor arguments.`

Always sketchy to refer to `this` in a constructor. I understand the restriction.